https://bugs.launchpad.net/ubuntu/+source/mtools/+bug/481482

From: Michael Casadevall <mcasadevall@ubuntu.com>

On karmic and lucid, the newer version of mtools introduces a
regression that causes filenames to break. According to the mtools
documentation (http://mtools.linux.lu/mtools.html#SEC7), long
filenames beyond eight characters should be truncated down to six
characters then a tile and number. i.e. thisisatest becomes
THISIS~1 for the short filename.

Signed-off-by: Anibal Monsalve Salazar <anibal@debian.org>

--- a/charsetConv.c	2009-03-01 09:35:14.000000000 +1100
+++ b/charsetConv.c	2010-02-15 14:19:53.000000000 +1100
@@ -410,7 +410,7 @@ int native_to_wchar(const char *native, 
 			break;
 		native += r;
 	}
-	if(mangled && end && native < end)
+	if(mangled && ((end && native < end) || (!end && *native &&  i == len)))
 		*mangled |= 3;
 	wchar[i]='\0';
 	return i;
--- a/file_name.c	2009-03-01 10:22:42.000000000 +1100
+++ b/file_name.c	2010-02-15 14:19:53.000000000 +1100
@@ -63,10 +63,6 @@ static void TranslateToDos(doscp_t *toDo
 
 	*Case = NONE;
 	for( ;  *s ; s++) {
-		if(!count) {
-			*mangled |= 3;
-			break;
-		}
 		/* skip spaces & dots */
 		if(*s == ' ' || *s == '.') {
 			*mangled |= 3;
@@ -91,7 +87,6 @@ static void TranslateToDos(doscp_t *toDo
 				*Case = UPPER;
 		} else
 			*t = *s;
-		count--;
 		t++;
 	}
 	wchar_to_dos(toDos, buffer, out, t - buffer, mangled);
